File: go-sdk/kestra_api_client/logs_api.go
package kestra_api_client

import (
	"context"
	"net/url"
	"os"
)

type LogsAPI struct {
	baseAPI
}

// logExecutionFilters builds the flat query params the per-execution log
// read/follow/download endpoints expect (minLevel, taskRunId, taskId, attempt).
func logExecutionFilters(minLevel, taskRunId, taskId *string, attempt *int) url.Values {
	return buildQueryParams("minLevel", minLevel, "taskRunId", taskRunId, "taskId", taskId, "attempt", attempt)
}

func (a *LogsAPI) ListLogsFromExecution(ctx context.Context, executionId, tenant string, minLevel, taskRunId, taskId *string, attempt *int) ([]LogEntry, error) {
	params := logExecutionFilters(minLevel, taskRunId, taskId, attempt)
	return doJSON[[]LogEntry](&a.baseAPI, ctx, "GET", tenantPath(tenant, "logs", executionId), nil, params)
}

// FollowLogsFromExecution opens an SSE stream that emits log entries for an execution.
// The returned channel is closed when the stream ends or the context is cancelled.
// The server opens the stream with a synthetic empty entry (id "start") so the SSE
// initializes even when there are no logs yet; callers filtering on entry fields
// should skip entries without an execution id.
func (a *LogsAPI) FollowLogsFromExecution(ctx context.Context, executionId, tenant string, minLevel *string) (<-chan *LogEntry, error) {
	path := tenantPath(tenant, "logs", executionId, "follow")
	params := logExecutionFilters(minLevel, nil, nil, nil)
	return followSSE[LogEntry](&a.baseAPI, ctx, path, params)
}

func (a *LogsAPI) DownloadLogsFromExecution(ctx context.Context, executionId, tenant string, minLevel, taskRunId, taskId *string, attempt *int) (*os.File, error) {
	params := logExecutionFilters(minLevel, taskRunId, taskId, attempt)
	return a.doDownload(ctx, "GET", tenantPath(tenant, "logs", executionId, "download"), nil, params, contentPlainText)
}

func (a *LogsAPI) DeleteLogsFromExecution(ctx context.Context, executionId, tenant string, minLevel, taskRunId, taskId *string, attempt *int) error {
	params := buildQueryParams("minLevel", minLevel, "taskRunId", taskRunId, "taskId", taskId, "attempt", attempt)
	return a.doVoidJSON(ctx, "DELETE", tenantPath(tenant, "logs", executionId), nil, params)
}

func (a *LogsAPI) DeleteLogsFromFlow(ctx context.Context, namespace, flowId, tenant string, triggerId *string) error {
	params := buildQueryParams("triggerId", triggerId)
	return a.doVoidJSON(ctx, "DELETE", tenantPath(tenant, "logs", namespace, flowId), nil, params)
}

func (a *LogsAPI) SearchLogs(ctx context.Context, tenant string, page, size *int, sort []string, filters []SearchFilter) (*PagedResultsLogEntry, error) {
	params := buildQueryParams("page", page, "size", size)
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[*PagedResultsLogEntry](&a.baseAPI, ctx, "GET", tenantPath(tenant, "logs", "search"), nil, params)
}

File: python/python-sdk/kestrapy/api/logs_api.py
from typing import Generator, List, Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.log_entry import LogEntry
from kestrapy.models.paged_results_log_entry import PagedResultsLogEntry
from kestrapy.models.query_filter import QueryFilter


class LogsApi(BaseApi):

    # ── Execution Logs ──────────────────────────────────────────────

    def list_logs_from_execution(
        self,
        execution_id: str,
        tenant: str,
        min_level: Optional[str] = None,
        task_run_id: Optional[str] = None,
        task_id: Optional[str] = None,
        attempt: Optional[int] = None,
    ) -> List[LogEntry]:
        path = self._tenant_path(tenant, "logs", execution_id)
        params = self._build_query_params(
            minLevel=min_level, taskRunId=task_run_id,
            taskId=task_id, attempt=attempt,
        )
        return self._json_list_request("GET", path, LogEntry, params=params)

    def download_logs_from_execution(
        self,
        execution_id: str,
        tenant: str,
        min_level: Optional[str] = None,
        task_run_id: Optional[str] = None,
        task_id: Optional[str] = None,
        attempt: Optional[int] = None,
    ) -> str:
        path = self._tenant_path(tenant, "logs", execution_id, "download")
        params = self._build_query_params(
            minLevel=min_level, taskRunId=task_run_id,
            taskId=task_id, attempt=attempt,
        )
        return self._text_request("GET", path, params=params, accept=self.TEXT)

    def delete_logs_from_execution(
        self,
        execution_id: str,
        tenant: str,
        min_level: Optional[str] = None,
        task_run_id: Optional[str] = None,
        task_id: Optional[str] = None,
        attempt: Optional[int] = None,
    ) -> None:
        path = self._tenant_path(tenant, "logs", execution_id)
        params = self._build_query_params(
            minLevel=min_level, taskRunId=task_run_id,
            taskId=task_id, attempt=attempt,
        )
        self._void_request("DELETE", path, params=params)

    def follow_logs_from_execution(
        self,
        execution_id: str,
        tenant: str,
        min_level: Optional[str] = None,
    ) -> Generator[LogEntry, None, None]:
        path = self._tenant_path(tenant, "logs", execution_id, "follow")
        params = self._build_query_params(minLevel=min_level)
        return self._sse_stream(path, LogEntry, params=params)

    def delete_logs_from_flow(
        self,
        namespace: str,
        flow_id: str,
        trigger_id: str,
        tenant: str,
    ) -> None:
        path = self._tenant_path(tenant, "logs", namespace, flow_id)
        params = self._build_query_params(triggerId=trigger_id)
        self._void_request("DELETE", path, params=params)

    # ── Search ──────────────────────────────────────────────────────

    def search_logs(
        self,
        tenant: str,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
    ) -> PagedResultsLogEntry:
        path = self._tenant_path(tenant, "logs", "search")
        params = list(self._build_query_params(page=page, size=size).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._json_request("GET", path, PagedResultsLogEntry, params=params)
